File: Sandboxes/gaze_control/src/gaze_control_python/dreamer_joint_publisher.py
#!/usr/bin/env python
import rospy
import time
import xml.dom.minidom
import math
from sensor_msgs.msg import JointState

def get_param(name, value=None):
    private = "~%s" % name
    if rospy.has_param(private):
        return rospy.get_param(private)
    elif rospy.has_param(name):
        return rospy.get_param(name)
    else:
        return value

class Custom_Joint_Publisher():
    def __init__(self):
        self.joint_list = []  # for maintaining the original order of the joints      
        self.free_joints = {}  
        self.zeros = get_param("zeros")
        self.dependent_joints = get_param("dependent_joints", {})
        self.load_joint_information()

        self.pub = rospy.Publisher('joint_states', JointState, queue_size=5)

    def load_joint_information(self):
        # Load Robot Description for ROS parameter server
        description = get_param('robot_description')
        robot = xml.dom.minidom.parseString(description).getElementsByTagName('robot')[0]
        # Find all non-fixed joints
        for child in robot.childNodes:
            if child.nodeType is child.TEXT_NODE:
                continue
            if child.localName == 'joint':
                jtype = child.getAttribute('type')
                if jtype == 'fixed' or jtype == 'floating':
                    continue
                name = child.getAttribute('name')
                self.joint_list.append(name)

                if jtype == 'continuous':
                    minval = -pi
                    maxval = pi
                else:
                    try:
                        limit = child.getElementsByTagName('limit')[0]
                        minval = float(limit.getAttribute('lower'))
                        maxval = float(limit.getAttribute('upper'))
                    except:
                        rospy.logwarn("%s is not fixed, nor continuous, but limits are not specified!" % name)
                        continue

                # -- Not useful for the head --
                # For mimic joints
                mimic_tags = child.getElementsByTagName('mimic')
                if len(mimic_tags) == 1:
                    tag = mimic_tags[0]
                    entry = {'parent': tag.getAttribute('joint')}
                    if tag.hasAttribute('multiplier'):
                        entry['factor'] = float(tag.getAttribute('multiplier'))
                    if tag.hasAttribute('offset'):
                        entry['offset'] = float(tag.getAttribute('offset'))

                    self.dependent_joints[name] = entry
                    continue

                if name in self.dependent_joints:
                    continue

                if self.zeros and name in self.zeros:
                    zeroval = self.zeros[name]
                elif minval > 0 or maxval < 0:
                    zeroval = (maxval + minval)/2
                else:
                    zeroval = 0

                # For each free joint, create with its zero starting position, min, max values, and if its continuous or not
                joint = {'min': minval, 'max': maxval, 'zero': zeroval}
                # Add and set position to 0
                joint['position'] = zeroval

                if jtype == 'continuous':
                    joint['continuous'] = True
                self.free_joints[name] = joint

        #self.print_joints()

    def print_joints(self):
        print "All Joints"
        for joint in self.joint_list:
            print joint
        print "Controllable Joints"
        for joint_name in self.joint_list:
            if joint_name not in self.dependent_joints:
                print joint_name

        #print self.free_joints

    def publish_joints(self):
        msg = JointState()
        msg.header.stamp = rospy.Time.now()

        for i, name in enumerate(self.joint_list):
            joint = None

            # Add Free Joint
            if name in self.free_joints:
                joint = self.free_joints[name]
                factor = 1
                offset = 0
            # Add Dependent Joint
            elif name in self.dependent_joints:
                param = self.dependent_joints[name]
                parent = param['parent']
                joint = self.free_joints[parent]
                factor = param.get('factor', 1)
                offset = param.get('offset', 0)

            joint_position = joint['position'] * factor + offset
            #print name, ":", joint['position']

            msg.name.append(str(name))
            msg.position.append(joint_position)
            self.pub.publish(msg)

File: Sandboxes/gaze_control/src/gaze_control_cpp/dreamerJointPublisher.cpp
/**
 * dreamerJointPublisher.cpp
 * Holds robot joint limits
 * Publishes robot joint list to RVIZ
 */

#include <iostream>
#include <cstring>
#include <cmath>
#include <vector>
#include <map>

#include "ros/ros.h"
#include <tinyxml.h>
#include <sensor_msgs/JointState.h>

#include "dreamerJointPublisher.h"

// Generic Constructor:
// Loads joints from external node
// Initializes publishing node
dreamerJointPublisher::dreamerJointPublisher(){
	loadJointInformation(jointNodeHandler);
	jointPublisher = jointNodeHandler.advertise<sensor_msgs::JointState>("joint_states", 5);
}

// Generic Destructor
dreamerJointPublisher::~dreamerJointPublisher(void){}


/**
 * Function: Load joint information from dreamer's URDF model
 * Inputs: ROS Node handle class
 * Returns: None
 */
void dreamerJointPublisher::loadJointInformation(ros::NodeHandle n){
	std::string retrieve;
	// Query for "robot_description" parameter
	if(jointNodeHandler.hasParam("robot_description")){
		// Place retrieved parameter in string
		jointNodeHandler.getParam("robot_description", retrieve);
		// std::cout << "Loaded Robot Description" << std::endl;

		// Declare xml parser
		TiXmlDocument robotXML;
		if(robotXML.Parse(retrieve.c_str(), 0, TIXML_ENCODING_UTF8)){

			// Pick out the first element in the xml
			TiXmlHandle hrob(&robotXML);
			TiXmlElement *ele = hrob.FirstChildElement("robot").FirstChildElement().ToElement();

			// Iterate through each element 
			for(ele; ele; ele = ele->NextSiblingElement()){
				// Only retrieve "joint" elements
				if(ele->ValueStr() == "joint"){
					std::string type = ele->Attribute("type");					
					std::string name = ele->Attribute("name");
					
					// Ignore fixed/floating joints
					if(type == "fixed" || type == "floating")
						continue;

					// Add joint to jointList vector
					jointList.push_back(name);

					double minval;
					double maxval;
					double zeroval;
					if(type == "continuous"){
						minval = -M_PI;
						maxval = M_PI;
					}
					else{
						try{
							TiXmlElement *limit = ele->FirstChildElement("limit");
							minval = atof(limit->Attribute("lower"));
							maxval = atof(limit->Attribute("upper"));
							// std::cout << "Retrieved " << name << " joint" << std::endl;
						}
						catch(...){
							std::cout << "Failed to get " << ele->Value() << " limits" << std::endl;
							continue;
						}
					}

					// Ignoring mimic tags
					std::string zeroes;
					if(jointNodeHandler.hasParam("zeros")){
						// Place retrieved parameter in string
						jointNodeHandler.getParam("zeros", zeroes);
						zeroval = 0; // TODO actually fill in correct zero value
					}
					else if(minval > 0 || maxval < 0)
						zeroval = ( maxval + minval ) / 2;
					else
						zeroval = 0;

					// Fill in map for joint information
					std::map<std::string, double> joint;
					joint["min"] = minval;
					joint["max"] = maxval;
					joint["zero"] = zeroval;
					joint["position"] = zeroval;

					// Continuous is True
					if(type == "continuous"){
						joint["continuous"] = 1;
					}

					// Add joint to free joint map
					freeJoints[name] = joint;
				}

			}

		}

		std::cout << "Dreamer Joints loaded" << std::endl;
	}

	else
		std::cout << "Failed to find robot parameters" << std::endl;
	
}


/**
 * Function: Send joint positions to RVIZ
 * Inputs: None
 * Returns: None
 */
void dreamerJointPublisher::publishJoints(void){	
	// std::cout << "Publishing joints" << std::endl;
	sensor_msgs::JointState msg;
	msg.header.stamp = ros::Time::now();

	for(int i = 0; i < jointList.size(); i++){
		if(freeJoints.find(jointList[i]) != freeJoints.end()){
			// std::cout << jointList[i] << std::endl;
			msg.name.push_back(jointList[i]);
			msg.position.push_back(freeJoints[jointList[i]]["position"]);
			// msg.position.push_back(.1);

		}
		else{
			msg.name.push_back(jointList[i]);
			msg.position.push_back(0);
		}
	 
	
	}
	// std::cout << msg << std::endl;
	jointPublisher.publish(msg);
	
}


/**
 * Function: Prints out each free joint and it corresponding values
 * Inputs: None
 * Returns: None
 */
void dreamerJointPublisher::printDebug(void){
	std::map < std::string, std::map<std::string, double> >::iterator it;
	for(it = freeJoints.begin(); it != freeJoints.end(); it++){
		std::cout << it->first << ":\t";
		std::cout << "min: " << it->second["min"];
		std::cout << "\tmax: " << it->second["max"];
		std::cout << "\tzero: " << it->second["zero"] << std::endl;
	}
}
